fix: Add new benchmark beRuSciBench along with AbsTaskTextRegression

AlexeyVatolin opened the pull request:
Add RuSciBench datasets with scientific tasks on Russian and English from Russian scientific electronic library elibrary.ru

Here is out paper:
https://link.springer.com/article/10.1134/S1064562424602191

### Checklist
<!-- please do not delete this checklist -->

- [x] I did not add a dataset, or if I did, I added the [dataset checklist](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_dataset.md#submit-a-pr) to the PR and completed it.
- [x] I did not add a model, or if I did, I added the [model checklist](https://github.com/embeddings-benchmark/mteb/blob/main/docs/adding_a_model.md#submitting-your-model-as-a-pr) to the PR and completed it.


- [x] I have run the following models on the task (adding the results to the pr). These can be run using the `mteb run -m {model_name} -t {task_name}` command.
  - [x] `sentence-transformers/paraphrase-multilingual-MiniLM-L12-v2`
  - [x] `intfloat/multilingual-e5-small`
- [x] I have checked that the performance is neither trivial (both models gain close to perfect scores) nor random (both models gain close to random scores).
- [x] I have considered the size of the dataset and reduced it if it is too big (2048 examples is typically large enough for most tasks)

| model_name                            | task_name                        | languages         |   main_score |
|:--------------------------------------|:---------------------------------|:------------------|-------------:|
| multilingual-e5-small                 | RuSciBenchBitexMining            | eng-Latn,rus-Cyrl |     0.978372 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchBitexMining            | eng-Latn,rus-Cyrl |     0.945861 |
| multilingual-e5-small                 | RuSciBenchBitexMining            | rus-Cyrl,eng-Latn |     0.974774 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchBitexMining            | rus-Cyrl,eng-Latn |     0.929254 |
| multilingual-e5-small                 | RuSciBenchCiteRetrieval          | eng-Latn          |     0.25836  |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCiteRetrieval          | eng-Latn          |     0.23692  |
| multilingual-e5-small                 | RuSciBenchCiteRetrieval          | rus-Cyrl          |     0.28923  |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCiteRetrieval          | rus-Cyrl          |     0.18175  |
| multilingual-e5-small                 | RuSciBenchCociteRetrieval        | eng-Latn          |     0.21956  |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCociteRetrieval        | eng-Latn          |     0.2035   |
| multilingual-e5-small                 | RuSciBenchCociteRetrieval        | rus-Cyrl          |     0.24766  |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCociteRetrieval        | rus-Cyrl          |     0.15751  |
| multilingual-e5-small                 | RuSciBenchCoreRiscClassification | eng-Latn          |     0.594057 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCoreRiscClassification | eng-Latn          |     0.578581 |
| multilingual-e5-small                 | RuSciBenchCoreRiscClassification | rus-Cyrl          |     0.594652 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCoreRiscClassification | rus-Cyrl          |     0.580301 |
| multilingual-e5-small                 | RuSciBenchPubTypeClassification  | eng-Latn          |     0.345671 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchPubTypeClassification  | eng-Latn          |     0.317749 |
| multilingual-e5-small                 | RuSciBenchPubTypeClassification  | rus-Cyrl          |     0.361472 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchPubTypeClassification  | rus-Cyrl          |     0.321645 |

Samoed commented on mteb/tasks/BitextMining/multilingual/RuSciBenchBitexMining.py:
```diff
@@ -0,0 +1,71 @@
+from mteb.abstasks.MultilingualTask import MultilingualTask
+from mteb.abstasks.TaskMetadata import TaskMetadata
+
+_LANGUAGES = {
+    "ru-en": ["rus-Cyrl", "eng-Latn"],
+    "en-ru": ["eng-Latn", "rus-Cyrl"],
+}
```
Do we need both configs?

Samoed commented on mteb/tasks/Retrieval/multilingual/RuSciBenchRetrieval.py:
```diff
@@ -0,0 +1,127 @@
+        if self.data_loaded:
+            return
+
+        self.corpus, self.queries, self.relevant_docs = load_neuclir_data(
```
Can you copy `load_neuclir_data` in this file?

Samoed reviewed (commented):
Congratulations on the publication of your paper!

KennethEnevoldsen commented on mteb/tasks/BitextMining/multilingual/RuSciBenchBitexMining.py:
```diff
@@ -0,0 +1,71 @@
+        prompt="Given the following title and abstract of the scientific article, find its translation",
+    )
+
+    def load_data(self, **kwargs):
+        if self.data_loaded:
+            return
+
+        self.dataset = {}
+        for lang in self.hf_subsets:
+            self.dataset.setdefault(lang, {})[_SPLITS[0]] = datasets.load_dataset(
+                split=_SPLITS[0],
+                name=lang,
+                **self.metadata_dict["dataset"],
+            )
+
+        self.dataset_transform()
+        self.data_loaded = True
```
I don't think it is needed to overwrite load data here, you can simply provide the dataset in the correct shape (you can always do this is a branch if you want to keep the structure in the original dataset)

KennethEnevoldsen commented on mteb/tasks/BitextMining/multilingual/RuSciBenchBitexMining.py:
```diff
@@ -0,0 +1,71 @@
+            "path": "mlsa-iai-msu-lab/ru_sci_bench_bitext_mining",
+            "revision": "927d95897a168b79568e96591276b995ed1c4da8",
+        },
+        description="Find translation of a scientific article",
```
I would love a more detailed description here. What scientific articles, how was the translation done etc.

KennethEnevoldsen commented on mteb/tasks/Classification/multilingual/RuSciBenchClassification.py:
```diff
@@ -0,0 +1,91 @@
+            "path": "mlsa-iai-msu-lab/ru_sci_bench_mteb",
+            "revision": "fbc0599a0b5f00b3c7d87ab4d13490f04fb77f8e",
+        },
+        description="Classification of scientific papers (title+abstract) by publication Core RISC status",
```
What is the publication core RISC status

KennethEnevoldsen commented on mteb/tasks/Classification/multilingual/RuSciBenchClassification.py:
```diff
@@ -0,0 +1,91 @@
+            "path": "mlsa-iai-msu-lab/ru_sci_bench_mteb",
+            "revision": "fbc0599a0b5f00b3c7d87ab4d13490f04fb77f8e",
+        },
+        description="Classification of scientific papers (title+abstract) by publication type",
```
I would love to know what the publication type is

KennethEnevoldsen commented on mteb/tasks/Retrieval/multilingual/RuSciBenchRetrieval.py:
```diff
@@ -0,0 +1,127 @@
+            "path": "mlsa-iai-msu-lab/ru_sci_bench_cite_retrieval",
+            "revision": "6cb447d02f41b8b775d5d9df7faf472f44d2f1db",
+        },
+        description="Retrieval of related scientific papers based on their title and abstract",
```
```suggestion
        description="Retrieval of related scientific papers based on their title and abstract.",
```

Good. Could potentially add some information on the paper's origin.

I would especially like to have enough information to distinguish these from the below

KennethEnevoldsen reviewed (commented):
Metadata generally looks good. Though maybe the descriptions could use a slight improvement

isaac-chung commented:
@AlexeyVatolin would love to get this in, if you're still working on this!

AlexeyVatolin commented:
As @Samoed mentioned, I have added RuSciBench to the list of benchmarks. There is an issue with the GRNTI and OECD classification tasks: they were previously added as part of the RuMTEB benchmark, but only in Russian. To avoid name conflicts, I added «Orig» to the names (RuSciBenchGRNTIOrigClassification, RuSciBenchOECDOrigClassification). I have checked and found that the data is sampled slightly differently, which is why the metric values for the tasks do not match in Russian.

isaac-chung commented:
Thanks. I think in general this looks good. I'd like get @KennethEnevoldsen and @Samoed 's opinion on the added regression abstask before moving forward. 

> I added "Orig" to the names (RuSciBenchGRNTIOrigClassification, RuSciBenchOECDOrigClassification)

Let's add `superseded_by` to the non-orig version of the tasks as well? e.g.
- add `superseded_by="RuSciBenchOECDOrigClassification"` to `RuSciBenchOECDClassification`

AlexeyVatolin commented on mteb/abstasks/AbsTaskRegression.py:
```diff
@@ -0,0 +1,171 @@
+    max_value: float
+
+
+class AbsTaskRegression(AbsTask):
```
Done

Samoed commented on mteb/abstasks/AbsTaskRegression.py:
```diff
@@ -0,0 +1,180 @@
+        eval_split = dataset[hf_split]
+
+        evaluator = self.evaluator(
+            train_split[self.input_column_name],
+            train_split[self.label_column_name],
```
Maybe downsample train dataset somehow? But I don't have ideas how to do this correctly

Samoed replied:
I think this would work

AlexeyVatolin replied:
I have added the stratified downsampling that I mentioned earlier

AlexeyVatolin commented on mteb/abstasks/AbsTaskRegression.py:
```diff
@@ -0,0 +1,229 @@
+        return scores
+
+    @staticmethod
+    def stratified_subsampling(
```
Ok, I added. Here is scores table for both regression tasks
| model_name                            | task_name                         | languages         |   main_score |
|:--------------------------------------|:----------------------------------|:------------------|-------------:|
| multilingual-e5-small                 | RuSciBenchCitedCountRegression    | eng-Latn          |     0.061745 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCitedCountRegression    | eng-Latn          |     0.056524 |
| multilingual-e5-small                 | RuSciBenchCitedCountRegression    | rus-Cyrl          |     0.07181  |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchCitedCountRegression    | rus-Cyrl          |     0.05654  |
| multilingual-e5-small                 | RuSciBenchYearPublRegression      | eng-Latn          |     0.284397 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchYearPublRegression      | eng-Latn          |     0.198798 |
| multilingual-e5-small                 | RuSciBenchYearPublRegression      | rus-Cyrl          |     0.244046 |
| paraphrase-multilingual-MiniLM-L12-v2 | RuSciBenchYearPublRegression      | rus-Cyrl          |     0.153153 |

KennethEnevoldsen commented on mteb/abstasks/AbsTaskRegression.py:
```diff
@@ -0,0 +1,269 @@
+    max_value: float
+
+
+class AbsTaskRegression(AbsTask):
```
Assuming it only worsens for text regression let us rename it to `AbsTaskTextRegression` (for v2 compatibility)

KennethEnevoldsen commented on mteb/evaluation/evaluators/RegressionEvaluator.py:
```diff
@@ -0,0 +1,85 @@
+            )
+            test_cache = X_test
+
+        linear_regression = LinearRegression(fit_intercept=self.fit_intercept)
```
Since people could use another regression model, then I would probably just allow people to pass in an sklearn regressor.

See an example from v2 [here](https://github.com/embeddings-benchmark/mteb/blob/ba7bb7df3d0e4fa1cbca60f1ee1eff1887186172/mteb/evaluation/evaluators/ClassificationEvaluator.py#L37C9-L37C19)

KennethEnevoldsen reviewed (commented):
Focused mostly on the regression tasks - generally everything looks good, but had a few minor changes to add.

KennethEnevoldsen commented on mteb/evaluation/evaluators/RegressionEvaluator.py:
```diff
@@ -0,0 +1,90 @@
+
+class RegressorModel(Protocol):
+    def fit(self, X, y, sample_weight=None): ...
+    def predict(self, X): ...
```
Ahh, that is annoying..., but I see the reason for using this approach then. Let's rename it to SklearnRegressorModel (just to clarify that it is a Sklearn compatible model)

KennethEnevoldsen reviewed (approved):
Once the issue with the regressor typing is fixed then this is good to merge

AlexeyVatolin commented:
@KennethEnevoldsen, Could you please take a look at the pull request when you have a moment? Thank you!

KennethEnevoldsen reviewed (approved):
Looks great! Sorry for being slow to respond, I was at conference (ACL) last week
